File: bun-civitai-browser-master/src/civitai-api/v1/models/model-id.ts
import { modelTypesSchema } from "./base-models/misc";
import { type } from "arktype";
import { creatorSchema } from "./creators";
import {
  modelFileSchema,
  modelImageSchema,
  modelVersionStatsSchema,
  modelStatsSchema,
} from "./shared-types";

/**
 * Model version schema for the model-id endpoint
 * Represents a version of a specific model
 */
export const modelByIdVersionSchema = type({
  id: "number.integer",
  index: "number.integer", // the position in modelId.modelVersions array.
  name: "string",
  baseModel: "string",
  baseModelType: "string | null",
  "publishedAt?": "string.date.iso", // ISO 8601
  availability: "'EarlyAccess' | 'Public'",
  nsfwLevel: "number.integer",
  "description?": "string | null", // html doc strings
  "trainedWords?": "string[]",
  stats: modelVersionStatsSchema,
  // covered: 'boolean', // have cover image or not
  files: modelFileSchema.array(),
  images: modelImageSchema.array(),
});
export type ModelByIdVersion = typeof modelByIdVersionSchema.infer;

/**
 * Model schema for the model-id endpoint
 * Represents a specific model with all its versions
 */
export const modelByIdSchema = type({
  id: "number.integer",
  name: "string",
  description: "string | null",
  // allowNoCredit: 'boolean',
  // allowCommercialUse: allowCommercialUse.array(),
  // allowDerivatives: 'boolean',
  // allowDifferentLicense: 'boolean',
  type: modelTypesSchema,
  poi: "boolean",
  nsfw: "boolean",
  nsfwLevel: "number.integer",
  // cosmetic: "null",
  "creator?": creatorSchema, // sometimes the user might deleted their account, left this field be null.
  stats: modelStatsSchema,
  tags: "string[]",
  modelVersions: modelByIdVersionSchema.array(),
});
export type ModelById = typeof modelByIdSchema.infer;

/**
 * Schema for tracking existing model versions on disk
 */
export const existedModelVersionsSchema = type({
  versionId: "number.integer",
  filesOnDisk: "number.integer[]",
}).array();
export type ExistedModelVersions = typeof existedModelVersionsSchema.infer;

File: lib/civitai_api/models/model_id.dart
import 'package:freezed_annotation/freezed_annotation.dart';

import 'creator.dart';
import 'shared.dart';

part 'model_id.freezed.dart';
part 'model_id.g.dart';

/// A model version as returned by the model-id endpoint.
/// (Images do NOT include an id field.)
@freezed
abstract class ModelByIdVersion with _$ModelByIdVersion {
  const factory ModelByIdVersion({
    required int id,
    required int index,
    required String name,
    required String baseModel,
    String? baseModelType,
    DateTime? publishedAt,
    @Default('Public') String availability,
    required int nsfwLevel,
    String? description,
    @Default([]) List<String> trainedWords,
    @Default(ModelVersionStats()) ModelVersionStats stats,
    @Default([]) List<ModelFile> files,
    @Default([]) List<ModelImage> images,
  }) = _ModelByIdVersion;

  factory ModelByIdVersion.fromJson(Map<String, dynamic> json) =>
      _$ModelByIdVersionFromJson(json);
}

/// A model as returned by GET /models/:id.
@freezed
abstract class ModelById with _$ModelById {
  const factory ModelById({
    required int id,
    required String name,
    String? description,
    @Default('Other') String type,
    @Default(false) bool poi,
    @Default(false) bool nsfw,
    required int nsfwLevel,
    Creator? creator,
    @Default(ModelStats()) ModelStats stats,
    @Default([]) List<String> tags,
    @Default([]) List<ModelByIdVersion> modelVersions,
  }) = _ModelById;

  factory ModelById.fromJson(Map<String, dynamic> json) =>
      _$ModelByIdFromJson(json);
}
